umbral/openssl.py

```diff
@@ -138,3 +143,23 @@ def _get_affine_coords_via_EC_POINT(ec_point, ec_group=None, curve_nid: int=None
         )
         backend.openssl_assert(res == 1)
     return (affine_x, affine_y)
+
+
+@contextmanager
+def _tmp_bn_mont_ctx(modulus):
+    """
+    Initializes and returns a BN_MONT_CTX for Montgomery ops.
+    Requires a modulus to place in the Montgomery structure.
+    """
+    bn_mont_ctx = backend._lib.BN_MONT_CTX_new()
+    backend.openssl_assert(bn_mont_ctx != backend._ffi.NULL)
+    # Don't set the garbage collector. Only free it when the context is done
+    # or else you'll get a null pointer error.
+
+    with backend._tmp_bn_ctx() as bn_ctx:
+        res = backend._lib.BN_MONT_CTX_set(bn_mont_ctx, modulus, bn_ctx)
+        backend.openssl_assert(res == 1)
+        try:
+            yield bn_mont_ctx
+        finally:
```

Comment:
Love this pattern - glad to see you adopting it!  :-)

I wonder, though - do we perhaps want the try-finally *outside* the with?  Currently, if there's an unhandled exception on any of the lines above (say, due to an incorrect OpenSSL or cryptography.io version), then the `bn_mont_ctx` will never be freed.

Reply:
Hmm, not sure..

If you notice that I set the BN_MONT_CTX with a modulus and also BN_CTX. My concern is that if we let the BN_CTX fall out of the context, the BN_MONT_CTX structure will fail.

I don't know how this works due to the mix between Python and C and how it would be handled, but I have a feeling that it must be used within the BN_CTX context.

Reply:
Hmm - I think you might be misunderstanding what I was saying.  I'm thinking this:

```
    try:
        with backend._tmp_bn_ctx() as bn_ctx:
            res = backend._lib.BN_MONT_CTX_set(bn_mont_ctx, modulus, bn_ctx)
            backend.openssl_assert(res == 1)
            yield bn_mont_ctx
    finally:
        backend._lib.BN_MONT_CTX_free(bn_mont_ctx)
```

...or are you saying that you have to be in the `bn_ctx` to *free* (but not init) the `bn_mont_ctx`?  I don't think that's so, although I haven't tried it.

Reply:
Oh, I see what you're saying. Yeah, you could do that.